patroni/config_generator.py

```diff
@@ -0,0 +1,466 @@
+"""patroni --generate-config machinery."""
+import abc
+import copy
+import os
+import psutil
+import socket
+import sys
+import yaml
+
+from getpass import getuser, getpass
+from contextlib import contextmanager
+from typing import Any, Dict, Iterator, List, Optional, Tuple, TYPE_CHECKING, Union
+if TYPE_CHECKING:  # pragma: no cover
+    from psycopg import Cursor
+    from psycopg2 import cursor
+
+from . import psycopg
+from .config import Config
+from .exceptions import PatroniException
+from .postgresql.config import parse_dsn
+from .postgresql.config import ConfigHandler
+from .postgresql.misc import postgres_major_version_to_int
+from .utils import get_major_version, patch_config, read_stripped
+
+
+# Mapping between the libpq connection parameters and the environment variables.
+# This dict should be kept in sync with `patroni.utils._AUTH_ALLOWED_PARAMETERS`
+# (we use "username" in the Patroni config for some reason, other parameter names are the same).
+_AUTH_ALLOWED_PARAMETERS_MAPPING = {
+    'user': 'PGUSER',
+    'password': 'PGPASSWORD',
+    'sslmode': 'PGSSLMODE',
+    'sslcert': 'PGSSLCERT',
+    'sslkey': 'PGSSLKEY',
+    'sslpassword': '',
+    'sslrootcert': 'PGSSLROOTCERT',
+    'sslcrl': 'PGSSLCRL',
+    'sslcrldir': 'PGSSLCRLDIR',
+    'gssencmode': 'PGGSSENCMODE',
+    'channel_binding': 'PGCHANNELBINDING'
+}
+_NO_VALUE_MSG = '#FIXME'
+
+
+def get_address() -> Tuple[str, str]:
+    """Try to get hostname and the ip address for it returned by :func:`~socket.gethostname`.
+
+    .. note::
+        Can also return local ip.
+
+    :returns: tuple consisting of the hostname returned by `~socket.gethostname`
+        and the first element in the sorted list of the addresses returned by :func:`~socket.getaddrinfo`.
+        Sorting guarantees it will prefer IPv4.
+
+    :raises:
+        :class:`PatroniException`: if :exc:`OSError` occured
+    """
+    hostname = None
+    try:
+        hostname = socket.gethostname()
+        return hostname, sorted(socket.getaddrinfo(hostname, 0, socket.AF_UNSPEC, socket.SOCK_STREAM, 0),
+                                key=lambda x: x[0])[0][4][0]
+    except OSError as err:
+        raise PatroniException(f'Failed to define ip address: {err}')
+
+
+class AbstractConfigGenerator(abc.ABC):
+    """Object representing the generated Patroni config."""
+
+    _HOSTNAME, _IP = get_address()
```

Comment:
If something is wrong get_address() will raise an exception that will not be handled in the generate_config() function: https://github.com/zalando/patroni/pull/2786/files#diff-cfee4a0506f1e512ce4c1fb59d7f7bac2f7924ad20e7d6a679044f4c5f4d6c88R465-R466

Reply:
Example:

```python
$ cat test.py 
class TestException(Exception):
    pass


def test():
    raise TestException


class Test:
    foo = test()


if __name__ == '__main__':
    print('main')

$ python test.py 
Traceback (most recent call last):
  File "test.py", line 9, in <module>
    class Test:
  File "test.py", line 11, in Test
    foo = test()
  File "test.py", line 6, in test
    raise TestException
__main__.TestException
```